tensorboard/plugins/mesh_visualizer/mesh_plugin.py

```diff
@@ -0,0 +1,253 @@
+    """Converts a TensorEvent into a JSON-compatible response.
+
+    Args:
+      event: TensorEvent object containing data in proto format.
+      content_type: enum plugin_data_pb2.MeshPluginData.ContentType value,
+        representing content type in TensorEvent.
+      data_shape: list of dimensions sizes of the tensor.
+      config: rendering scene configuration as dictionary.
+    Returns:
+      Dictionary of transformed metadata.
+    """
+    return {
+        'wall_time': event.wall_time,
+        'step': event.step,
+        'content_type': content_type,
+        'config': config,
+        'data_shape': list(data_shape)
+    }
+
+  def _get_tensor_data(self, event, sample):
+    """Convert a TensorEvent into a JSON-compatible response."""
+    data = self._get_sample(event, sample)
+    return data
+
+  def _collect_tensor_events(self, request):
+    """Collects list of tensor events based on request."""
+    run = request.args.get('run')
+    tag = request.args.get('tag')
+
+    # TODO(b/128995556): investigate why this additional metadata mapping is
+    # necessary, it must have something todo with the lifecycle of the request.
+    # Make sure we populate tags mapping structures.
+    self.prepare_metadata()
+
+    # We fetch all the tensor events that contain tag.
+    tensor_events = []  # List of tuples (meta, tensor).
+    for instance_tag in self._tag_to_instance_tags[tag]:
+      tensors = self._multiplexer.Tensors(run, instance_tag)
+      meta = self._instance_tag_to_metadata[instance_tag]
+      tensor_events += [(meta, tensor) for tensor in tensors]
+
+    # Make sure tensors sorted by timestamp in ascending order.
+    tensor_events = sorted(
+        tensor_events, key=lambda tensor_data: tensor_data[1].wall_time)
+
+    return tensor_events
+
+  @wrappers.Request.application
+  def _serve_mesh_data(self, request):
+    """A route that returns data for particular summary of specified type.
+
+    Data can represent vertices coordinates, vertices indices in faces,
+    vertices colors and so on. Each mesh may have different combination of
+    abovementioned data and each type/part of mesh summary must be served as
+    separate roundtrip to the server.
+
+    Args:
+      request: werkzeug.Request containing content_type as a name of enum
+        plugin_data_pb2.MeshPluginData.ContentType.
+    Returns:
+      werkzeug.Response either float32 or int32 data in binary format.
+    """
+    tensor_events = self._collect_tensor_events(request)
+    content_type = request.args.get('content_type')
+    content_type = plugin_data_pb2.MeshPluginData.ContentType.Value(
+        content_type)
+    sample = int(request.args.get('sample', 0))
+
+    response = [
+        self._get_tensor_data(tensor, sample)
+        for meta, tensor in tensor_events
+        if meta.content_type == content_type
+    ]
+
+    np_type = np.float32
+    if content_type == plugin_data_pb2.MeshPluginData.ContentType.Value('FACE'):
```

Comment:
Can’t we just use `plugin_data_pb2.MeshPluginData.FACE` here? It’s
shorter, faster at runtime, in principle more statically analyzable, and
doesn’t overflow the line length guideline.

Reply:
This is what I was using, but for some reason unit tests in PY3 were failing on github because of that. Not sure why.